feat: allow object arguments in component tree to be inspected

SYU15 opened the pull request:
## Description
Checks to see if an argument passed into component tree is an object, if so, allows the user to click on the argument directly to inspect the object properties.

Since the entire row is clickable in addition to the argument, I've created a few different options for UX which are shown below. I reused existing styles/color palettes but definitely open to suggestions.

**Currently working on tests.** Was curious to know why there weren't integration tests in this mp (unless I am missing something). Was thinking about adding a couple for `component-tree-arg` if that isn't an anti-pattern.

## Screenshots

### **1. Most subtle - args is underlined when a user hovers over it**
![underline_args](https://user-images.githubusercontent.com/10149117/79703694-396be680-8262-11ea-8d11-25c204ece11e.gif)

### **2. Less subtle - args is underlined and highlighted when user hovers over**
![highlight_args2](https://user-images.githubusercontent.com/10149117/79703810-e5adcd00-8262-11ea-9450-5e13b209823c.gif)

### **3. Most apparent - args is a button that is clickable** 
![Button_args](https://user-images.githubusercontent.com/10149117/79703704-45f03f00-8262-11ea-8503-9879d6201d69.gif)

Addresses #1169 related to issue #1127

nummi commented:
Nice! So many options. We could even combine these and have a subtle version of 3 but only on hover.

SYU15 commented:
> Nice! So many options. We could even combine these and have a subtle version of 3 but only on hover.

That is doable! I assume the suggestion is a button that highlights on hover @nummi?

rwwagner90 commented:
Thanks for working on this @SYU15! I think I like the more subtle options, but I defer to @nummi on style decisions. Since we are making the clicks more explicit now, rather than just the whole row, perhaps we should do the same with the component itself? i.e. click the component name rather than anywhere in the row to inspect the component args.

I'm not sure if we want to change that or not, just throwing ideas out.

SYU15 commented:
> i.e. click the component name rather than anywhere in the row to inspect the component args.

How would the user know the component name was the clickable area? I guess I'm slightly more in favor of having the whole row clickable because it seems more intuitive/obvious that way but open to suggestions.

rwwagner90 commented:
@SYU15 yeah, it can stay as is. @nummi @chancancode any thoughts on which options you like best? 1,2 or 3? I like the more subtle ones, 1 or 2, but happy to go with everyone likes best.

chancancode commented:
Personally I am fine with any of these but wit a leaning towpaths not cluttering the screen when not hovering. So either 1, 2, what @nummi said, or showing all the click areas on the same row when the row itself is hovered.

When I raised the potential issues, I was not primarily worried about discoverability (although it certainly is an issue and these are good solutions). I was mostly worried about nesting clickable areas, which makes it easy to misclick something else by narrowly missing a pixel or two. Seems like it’s not as much of an issue as I worried about. I think the hover solutions here also helps (in addition to aiding discoverability) since you can physically see which of the two things will happen when you click.

We may also want to consider making the {{ and }} part of the click area, but exclude them from the visual highlight? (So you can hover on them to trigger the highlight, or click on them to inspect the arg, but it wouldn’t itself light up - I assume that doesn’t look good visually.)

Anyway i am pretty happy with any of these and we can also iterate a bit over time. Assuming it works, we can always start with the more basic/subtle options and add to it based on feedback after releasing.

rwwagner90 commented:
@SYU15 I vote we go with solution number 2 for now. I think it's a good middle ground, and we can always tweak it later.

chancancode commented on app/components/component-tree-arg.hbs:
```diff
@@ -0,0 +1,13 @@
+{{#if this.valueIsObject}}
```
```suggestion
{{#if this.isObject}}
```

chancancode commented on app/components/component-tree-arg.hbs:
```diff
@@ -0,0 +1,13 @@
+{{#if this.valueIsObject}}
+  <span
+    class="value-token arg-value-token"
+    {{on "click" this.inspectArgumentValue}}
```
```suggestion
    {{on "click" this.inspect}}
```

chancancode commented on app/components/component-tree-arg.hbs:
```diff
@@ -0,0 +1,13 @@
+{{#if this.valueIsObject}}
+  <span
```
```suggestion
  <button type="button"
```

chancancode commented on app/components/component-tree-arg.hbs:
```diff
@@ -0,0 +1,13 @@
+    {{on "click" this.inspectArgumentValue}}
+    data-test-arg-object
+  >
+    {{this.componentArgumentDisplay}}
```
```suggestion
    {{this.displayValue}}
```

chancancode commented on app/components/component-tree-arg.js:
```diff
@@ -0,0 +1,30 @@
+import { action } from '@ember/object';
+import { inject as service } from '@ember/service';
+
+export default class ComponentTreeArgs extends Component {
```
```suggestion
export default class ComponentTreeArg extends Component {
```

chancancode commented on app/components/component-tree-arg.js:
```diff
@@ -0,0 +1,30 @@
+export default class ComponentTreeArgs extends Component {
+  @service port;
+
+  get valueIsObject() {
```
```suggestion
  get isObject() {
```

chancancode commented on app/components/component-tree-arg.js:
```diff
@@ -0,0 +1,30 @@
+  @service port;
+
+  get valueIsObject() {
+    return typeof this.args.componentArgValue === 'object' && this.args.componentArgValue !== null;
```
```suggestion
    return typeof this.args.value === 'object' && this.args.value !== null;
```

chancancode commented on app/components/component-tree-arg.js:
```diff
@@ -0,0 +1,30 @@
+    return typeof this.args.componentArgValue === 'object' && this.args.componentArgValue !== null;
+  }
+
+  get componentArgumentDisplay() {
```
```suggestion
  get displayValue() {
```

chancancode commented on app/components/component-tree-item.hbs:
```diff
@@ -44,9 +44,7 @@
-              <span class="value-token">
-                {{component-argument value}}
-              </span>
+              <ComponentTreeArg @componentArgValue={{value}} />
```
```suggestion
              <ComponentTreeArg @value={{value}} />
```

SYU15 commented:
<img width="692" alt="Screen Shot 2020-04-27 at 7 17 43 PM" src="https://user-images.githubusercontent.com/10149117/80440034-b6c7d480-88bc-11ea-9327-689a4498fe88.png">

This is what it looks like now (imagine a pointer cursor as well, which the screenshot didn't capture). Have addressed the feedback from @chancancode.

nummi commented:
How does it feel with a border-radius on the highlight? Maybe the util class `rounded` would work (4px).

chancancode commented on tests/acceptance/component-tree-test.js:
```diff
@@ -354,4 +359,16 @@ module('Component Tab', function (hooks) {
+
+  test('Can inspect component arguments that are objects in component tree', async function (assert) {
+    await visit('/component-tree');
+
```
```suggestion

```

(extra new line)

SYU15 commented:
> How does it feel with a border-radius on the highlight? Maybe the util class `rounded` would work (4px).
> 
> Also, how does it look with the light theme?

Have added the border radius and I think it looks good, thanks for the suggestion!

The color scheme @nummi isn't perfect but it was kind of challenging to find a color that worked 
 decently under all conditions (which I have added below). Let me know if you can think of something more appropriate.

### Chrome Light:
<img width="534" alt="Screen Shot 2020-04-27 at 11 43 26 PM" src="https://user-images.githubusercontent.com/10149117/80456232-1ab0c400-88e2-11ea-8f73-a19b82e545e8.png">
____________________________________________________________________
<img width="626" alt="Screen Shot 2020-04-27 at 11 43 02 PM" src="https://user-images.githubusercontent.com/10149117/80456239-1c7a8780-88e2-11ea-8653-546287bb48af.png">

### Chrome Dark:
<img width="898" alt="Screen Shot 2020-04-27 at 11 43 49 PM" src="https://user-images.githubusercontent.com/10149117/80456237-1be1f100-88e2-11ea-97b2-b9fdfce5742d.png">
____________________________________________________________________
<img width="529" alt="Screen Shot 2020-04-27 at 11 43 58 PM" src="https://user-images.githubusercontent.com/10149117/80456243-1d131e00-88e2-11ea-91ee-e36cc53b4dca.png">

### Firefox:
<img width="520" alt="Screen Shot 2020-04-27 at 11 44 21 PM" src="https://user-images.githubusercontent.com/10149117/80456272-27cdb300-88e2-11ea-8a8a-56a1cdc05fb4.png">
____________________________________________________________________

<img width="543" alt="Screen Shot 2020-04-27 at 11 45 04 PM" src="https://user-images.githubusercontent.com/10149117/80456275-28fee000-88e2-11ea-8741-a63dc8a7edba.png">

chancancode commented on app/styles/component_tree.scss:
```diff
@@ -101,6 +101,18 @@
+
+  .arg-value-token:hover {
+    background-color: var(--spec03);
+    color: var(--white);
```
Instead of using a hardcoded color, try using one of the `--base` or `--spec` colors here (e.g. `--base00`), that way you will avoid the white-on-yellow low contrast problems when switching between light and dark themes. See https://github.com/emberjs/ember-inspector/blob/master/app/styles/colors.scss @nummi may have more concrete suggestion on which combination to pick.

chancancode replied:
The white-on-yellow contrast issue is probably a blocker, but should be an easy fix, it may or may not be the best looking option but switching to `--base00` should at least resolve the contrast problem.

chancancode commented on app/components/component-tree-item.hbs:
```diff
@@ -44,9 +44,7 @@
           {{#each @item.args.positional as |value|}}
             <div class="arg-token flex ml-2">
               {{if (is-string value) "\""}}
```
We can save this refactor for another PR since this looks almost ready to go, but it had occurred to me that we should probably move the surrounding quotes and mustache (optional, via a `@mustache={{true}}` argument) into the component template, and turn `isString` into a getter instead of a global helper. Just wanted to note this somewhere so we don't forget.

SYU15 replied:
Sounds good @chancancode

chancancode commented:
Not nearly important enough the hold up the PR, but if you are making anymore style tweaks, but I am guessing the `...` case (and only the `...` case, not strings, etc) may look better with a tiny bit of horizontal padding (somewhere between 2-4px each side) so the highlight doesn't hug the text too tightly. Not a big deal if you don't get to it or if it doesn't end up working out though.

SYU15 commented:
Thanks for the tips @chancancode. Looks much better now:

<img width="500" alt="Screen Shot 2020-04-28 at 5 25 42 PM" src="https://user-images.githubusercontent.com/10149117/80550608-7e380180-8975-11ea-8558-9d66f89ead73.png">
<img width="700" alt="Screen Shot 2020-04-28 at 5 25 19 PM" src="https://user-images.githubusercontent.com/10149117/80550615-7f692e80-8975-11ea-842b-3de487d4714a.png">
